_posts/2025-02-13-rfprs.md

```diff
@@ -0,0 +1,347 @@
+
+## PR descriptions
+
+The PR description sets the stage for the reviewer. Since they are going to review the PR
+commit-by-commit, the description should give them as much context as possible before they start.
+Assume that the reviewer does not have the context that you do.
+
+The description should roughly include these components:
+
+1. Links to tickets or other prior art.
+2. Overview of the problem or business goal.
+3. Description of your approach.
+4. Callouts to any parts which are complex or about which you are hoping for extra feedback.
+5. Edge cases and risks.
+
+It's worth noting that PRs which are small in scope will have a correspondingly smaller PR
+description. PRs which are large or complex will have a larger PR description. This is largely
+subjective, and at first I think it might feel like it's unnecessary work, but that is temporary. In
+fact I've found that by writing a detailed description, I will sometimes remember things I had
+forgotten to address during development. By writing the description *for the reviewer* this causes
+you to think about the things that the reviewer would probably have questions about. Anything you
+catch at this phase is a major win, as it is a short amount of time compared to the revisions to the
+PR which would otherwise follow.
+
+Code snippets (for example, to showcase a new interface you are adding) are often really helpful in
+the description.
+
+## Commits
+
+The reviewer should expect to review your PR in commit order. What order should the commits be in
+and what should the commits contain? Remember that we are assuming that the reviewer is coming in
+fresh and probably does not have the familiarity with the code that you do.
+
+In general I try to isolate what I consider to be "high value" and "low value" changes in separate
+commits. A high value commit would be something like adding a new trait, or implementing that
+trait for a number of domain models. Low value commits would be things like formatting, or code
+imports. To make it as easy as possible on the reviewer, allow them to briefly consider low value
+changes and spend more time on the high value changes.
+
+> Do not mix high value and low value changes in the same commits.
+
+Since the reviewer is reading in commit order, we should also strive to limit the scope of any
+particular commit. If your PR is adding and implementing five interfaces to all of the domain
+models, a reasonable approach would be to have each interface, and the implementation of that
+interface, be in its own commit. If you're sweating now because you're not comfortable with Git,
+don't worry -- we'll talk later about tools you can use to build these kinds of commits.
+
+Guidelines for good commits:
+
+1. Commits should be limited in scope.
+2. Do not mix high value and low value concerns in the same commit.
+3. Commit message titles should be a sentence that describes an action.
+4. Commit messages should end with a period.
```

Comment:
I don't know that point 4 adds anything that isn't covered by 3. Point 3 captures the spirit by saying commit messages should be sentences and your examples below drive that home by including punctuation.

What's the intent? If it's to make it clear that commit messages should be descriptive and not lazy shorthand, what about saying in 3 that a commit title should be a "full sentence"?

Personally, my commit message are more like PR titles than full sentences with punctuation. (I don't know the name for this style. It's somewhere between a newspaper article title without title case and a sentence.) I'm not opposed to using a period in commit message, but I don't use them. If I was on your team I'd have no problem falling in line. 😄

Reply:
Yeah, good point. I don't think point 4 is worth keeping. I yoinked it out and added your suggestion about "full sentence".